File: app/src/main/java/com/call/colorscreen/ledflash/service/NotificationService.java
package com.call.colorscreen.ledflash.service;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Build;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.telecom.TelecomManager;
import android.util.Log;

import androidx.annotation.RequiresApi;


import com.call.colorscreen.ledflash.util.PhoneUtils;

import java.lang.ref.WeakReference;

@SuppressLint("OverrideAbstract")
public class NotificationService extends NotificationListenerService {
    private static WeakReference<NotificationService> serviceWeakReference;
    private StatusBarNotification inCallNotification;
    private boolean isListen = false;

    public StatusBarNotification getInCallNotification() {
        return this.inCallNotification;
    }

    public static NotificationService get() {

        WeakReference<NotificationService> weakReference = serviceWeakReference;
        Log.e("TAN", "get: serviceWeakReference"+weakReference+"##");
        if (weakReference == null || weakReference.get() == null) {
            return null;
        }
        return serviceWeakReference.get();
    }

    public StatusBarNotification[] getActiveNotifications() {
        return super.getActiveNotifications();
    }

    public void onCreate() {
        super.onCreate();
        Log.e("TAN", "serviceWeakReference onCreate: ");
        serviceWeakReference = new WeakReference<>(this);
    }

    @Override
    public void onNotificationPosted(final StatusBarNotification statusBarNotification) {
        super.onNotificationPosted(statusBarNotification);
        Log.e("TAN", "onNotificationPosted: "+statusBarNotification);
        new Thread() {
            public void run() {
                super.run();
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    if (!statusBarNotification.getPackageName().contains("incallui") && !statusBarNotification.getPackageName().contains("dialer"))
                        return;
                    String str = "";
                    TelecomManager telecomManager = getApplicationContext().getSystemService(TelecomManager.class);
                    if (telecomManager != null) {
                        str = telecomManager.getDefaultDialerPackage() + "";
                    }
                    if (!statusBarNotification.getPackageName().contains("incallui") && !statusBarNotification.getPackageName().equals(str)) {
                        return;
                    }
                    if (NotificationService.this.isListen) {
                        new Thread() {
                            public void run() {
                                super.run();
                                PhoneUtils.get(getApplicationContext()).getPhoneFromNotificationListen(statusBarNotification);
                            }
                        }.start();
                        return;
                    }
                    NotificationService.this.inCallNotification = statusBarNotification;
                }
            }
        }.start();
    }

    public void stopListenColorCall() {
        getApplicationContext().stopService(new Intent(getApplicationContext(), NotificationService.class));
        this.isListen = false;
    }

    @RequiresApi(api = 28)
    public void startListenColorCall() {
        this.isListen = true;
    }
}

File: app/src/main/java/com/call/colorscreen/ledflash/service/NotificationService.kt
package com.call.colorscreen.ledflash.service

import android.content.Intent
import android.os.Build
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.telecom.TelecomManager
import androidx.annotation.RequiresApi
import com.call.colorscreen.ledflash.util.PhoneUtils
import java.lang.ref.WeakReference


class NotificationService : NotificationListenerService() {

    private var inCallNotification: StatusBarNotification? = null
    private var isListen = false

    fun getInCallNotification(): StatusBarNotification? {
        return inCallNotification
    }
    companion object{
        private var serviceWeakReference: WeakReference<NotificationService?>? = null
        fun get(): NotificationService? {
            val weakReference = serviceWeakReference
            return if (weakReference?.get() == null) {
                null
            } else serviceWeakReference!!.get()
        }
    }

    override fun onCreate() {
        super.onCreate()
        serviceWeakReference = WeakReference(this)
    }

    override fun onNotificationPosted(statusBarNotification: StatusBarNotification) {
        super.onNotificationPosted(statusBarNotification)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            if (!statusBarNotification.packageName.contains("incallui") && !statusBarNotification.packageName.contains(
                    "dialer"
                )
            ) return
            var str = ""
            val telecomManager = applicationContext.getSystemService(
                TelecomManager::class.java
            )
            if (telecomManager != null) {
                str = telecomManager.defaultDialerPackage + ""
            }
            if (!statusBarNotification.packageName.contains("incallui") && statusBarNotification.packageName != str) {
                return
            }
            if (isListen) {
                object : Thread() {
                    override fun run() {
                        super.run()
                        PhoneUtils.get(applicationContext)?.getPhoneFromNotificationListen(statusBarNotification)
                    }
                }.start()
                return
            }
            inCallNotification = statusBarNotification
        }
    }

    fun stopListenColorCall() {
        applicationContext.stopService(Intent(applicationContext, NotificationService::class.java))
        isListen = false
    }

    @RequiresApi(api = 28)
    fun startListenColorCall() {
        isListen = true
    }
}